[Index] Record the access level of declarations in the index

ahoppen opened the pull request:
Depends on https://github.com/swiftlang/llvm-project/pull/11716

---

While updating the test cases, I noticed two slight oddities, for which I am happy with the current behavior, but I’m happy to be convinced by a different opinion.
- Function parameters are marked as `fileprivate`. I think we could make an argument that they should be marked as `local` instead of `fileprivate` even though they are indexed without index-locals enabled.
- Accessors such as `willSet` are marked as `fileprivate`. Since you can’t call them from anywhere else, I think this makes as much sense as anything. At least it matches the compiler’s effective access level.

rdar://163256878

ahoppen commented:
https://github.com/swiftlang/llvm-project/pull/11716

@swift-ci Please smoke test

bnbarham commented on lib/Index/IndexSymbol.cpp:
```diff
@@ -254,6 +254,30 @@ SymbolInfo index::getSymbolInfoForDecl(const Decl *D) {
   if (isLocalSymbol(D)) {
     info.Properties |= SymbolProperty::Local;
+  } else if (auto *VD = dyn_cast<ValueDecl>(D)) {
+    switch (VD->getEffectiveAccess()) {
```
Access level was added to the index request (not the index itself) in https://github.com/swiftlang/swift/pull/31942 and there was a long discussion there about whether it should be effective access or `getFormalAccessScope`, with the conclusion being `getFormalAccessScope`.

One notable case there was:
```
private enum E {
    case a, b, c
}
extension E {
    private func f() { print("Hello") }
}
```

Where apparently `getEffectiveAccess` gives `fileprivate` for `f` (which it isn't).

bnbarham reviewed (commented):
> * Function parameters are marked as fileprivate. I think we could make an argument that they should be marked as local instead of fileprivate even though they are indexed without index-locals enabled.
> * Accessors such as willSet are marked as fileprivate. Since you can’t call them from anywhere else, I think this makes as much sense as anything. At least it matches the compiler’s effective access level.

My preference here would be to not include any access level for them.

Also CC @keith as IIRC you were thinking about adding access levels sometime in the past.

ahoppen committed:
[Index] Call `updateInfo` while constructing index symbols

This will allow us to inspect the updated symbol info to decided whether we should report access levels for a declaration.

ahoppen committed:
[Index] Record the access level of declarations in the index

rdar://163256878

ahoppen commented:
https://github.com/swiftlang/llvm-project/pull/11716

@swift-ci Please smoke test
